subsys/mgmt/osdp/src/osdp_pd.c

```diff
@@ -174,125 +260,113 @@ static int pd_decode_command(struct osdp_pd *pd, uint8_t *buf, int len)
-		osdp_cmd_enqueue(pd, cmd);
+		cmd.led.permanent.control_code = buf[pos++];
+		cmd.led.permanent.on_count = buf[pos++];
+		cmd.led.permanent.off_count = buf[pos++];
+		cmd.led.permanent.on_color = buf[pos++];
+		cmd.led.permanent.off_color = buf[pos++];
+		if (!do_command_callback(pd, &cmd)) {
+			break;
+		}
 		pd->reply_id = REPLY_ACK;
 		ret = OSDP_PD_ERR_NONE;
 		break;
 	case CMD_BUZ:
-		if (len != CMD_BUZ_DATA_LEN) {
+		if (len != CMD_BUZ_DATA_LEN || !pd->command_callback) {
 			break;
 		}
-		cmd = osdp_cmd_alloc(pd);
-		if (cmd == NULL) {
-			LOG_ERR("cmd alloc error");
+		cmd.id = OSDP_CMD_BUZZER;
+		cmd.buzzer.reader = buf[pos++];
+		cmd.buzzer.control_code = buf[pos++];
+		cmd.buzzer.on_count = buf[pos++];
+		cmd.buzzer.off_count = buf[pos++];
+		cmd.buzzer.rep_count = buf[pos++];
+		if (!do_command_callback(pd, &cmd)) {
 			break;
 		}
-		cmd->id = OSDP_CMD_BUZZER;
-		cmd->buzzer.reader       = buf[pos++];
-		cmd->buzzer.control_code = buf[pos++];
-		cmd->buzzer.on_count     = buf[pos++];
-		cmd->buzzer.off_count    = buf[pos++];
-		cmd->buzzer.rep_count    = buf[pos++];
-		osdp_cmd_enqueue(pd, cmd);
 		pd->reply_id = REPLY_ACK;
 		ret = OSDP_PD_ERR_NONE;
 		break;
 	case CMD_TEXT:
-		if (len < CMD_TEXT_DATA_LEN) {
+		if (len < CMD_TEXT_DATA_LEN || !pd->command_callback) {
 			break;
 		}
-		cmd = osdp_cmd_alloc(pd);
-		if (cmd == NULL) {
-			LOG_ERR("cmd alloc error");
+		cmd.id = OSDP_CMD_TEXT;
+		cmd.text.reader = buf[pos++];
+		cmd.text.control_code = buf[pos++];
+		cmd.text.temp_time = buf[pos++];
+		cmd.text.offset_row = buf[pos++];
+		cmd.text.offset_col = buf[pos++];
+		cmd.text.length = buf[pos++];
+		if (cmd.text.length > OSDP_CMD_TEXT_MAX_LEN ||
+		    ((len - CMD_TEXT_DATA_LEN) < cmd.text.length) ||
+		    cmd.text.length > OSDP_CMD_TEXT_MAX_LEN) {
 			break;
 		}
-		cmd->id = OSDP_CMD_TEXT;
-		cmd->text.reader       = buf[pos++];
-		cmd->text.control_code = buf[pos++];
-		cmd->text.temp_time    = buf[pos++];
-		cmd->text.offset_row   = buf[pos++];
-		cmd->text.offset_col   = buf[pos++];
-		cmd->text.length       = buf[pos++];
-		if (cmd->text.length > OSDP_CMD_TEXT_MAX_LEN ||
-		    ((len - CMD_TEXT_DATA_LEN) < cmd->text.length) ||
-		    cmd->text.length > OSDP_CMD_TEXT_MAX_LEN) {
-			osdp_cmd_free(pd, cmd);
-			break;
+		for (i = 0; i < cmd.text.length; i++) {
+			cmd.text.data[i] = buf[pos++];
 		}
-		for (i = 0; i < cmd->text.length; i++) {
-			cmd->text.data[i] = buf[pos++];
+		if (!do_command_callback(pd, &cmd)) {
+			break;
 		}
-		osdp_cmd_enqueue(pd, cmd);
 		pd->reply_id = REPLY_ACK;
 		ret = OSDP_PD_ERR_NONE;
 		break;
 	case CMD_COMSET:
-		if (len != CMD_COMSET_DATA_LEN) {
+		if (len != CMD_COMSET_DATA_LEN || !pd->command_callback) {
+			break;
+		}
+		cmd.id = OSDP_CMD_COMSET;
+		cmd.comset.address = buf[pos++];
+		cmd.comset.baud_rate = buf[pos++];
+		cmd.comset.baud_rate |= buf[pos++] << 8;
+		cmd.comset.baud_rate |= buf[pos++] << 16;
+		cmd.comset.baud_rate |= buf[pos++] << 24;
```

Comment:
nitpick: you can use `byteorder.h` API here.

Reply:
Thanks, will take a look at this in a subsequent PR.

Reply:
unresolved

Reply:
@gmarull, I wasn't planning on taking this up now (as mentioned earlier). There are other places in `subsys/mgmt/` where such a change needs to be made and it doesn't make sense to overload this PR any further.